idl/on_off.matter

```diff
@@ -0,0 +1,70 @@
+
```

Comment:
I am imagining we want to consider how to organise these as a separate problem. The definition itself does exist, however due the way code generation works (these files are a human readable representation of some "zap" input files) they are currently one single large file: https://github.com/project-chip/connectedhomeip/blob/master/src/controller/data_model/controller-clusters.matter

Ideally I would love them to be separated because to me it makes sense to only import what you need. Separating in the main repo is slightly more complex due to our zap dependency. We can also separate in our macros - parse the entire file and then pick what we codegen, however that seems painful (one large file parse, potentially many times).

Apart from that, I wonder also how we manage compatibility - if new things appear in the IDL, it may take a bit of time to add it to rust support, so we have to figure out a way to pin the version. 

My current thinking is that I agree with you that checking in these (and maintaining them) is annoying however I also have no good solution. My concerns to solve:
- how can I split out the "one large IDL" into individual pieces (or do we want one large file and pick from it)?
- how can I ensure compatibility so changes in CPP does not break matter-rs?
- feel uneasy about anything requiring network during the build - I would want some checkout and potentially submodules to be sufficient for compilation (some countries and setups have firewalls preventing generall internet)
- idl is the friendliest format we have, however there is also movement to generate xml directly out of spec text, which would be the most accurate. Those are split (https://github.com/project-chip/connectedhomeip/tree/master/data_model/clusters) however not currently fully correct/usable so we may want to hold off on that path (hence the current model includes a idl parser)

Reply:
Some answers inline:

> Ideally I would love them to be separated because to me it makes sense to only import what you need. Separating in the main repo is slightly more complex due to our zap dependency. We can also separate in our macros - parse the entire file and then pick what we codegen, however that seems painful (one large file parse, potentially many times).

Though and as you say parsing/code-generating takes (a few tens of) milliseconds only, so multiple parsing perhaps no big deal?

But then, see below.

> Apart from that, I wonder also how we manage compatibility - if new things appear in the IDL, it may take a bit of time to add it to rust support, so we have to figure out a way to pin the version.

Assuming IDL-as-one-large-file would stay like that for a while, perhaps we can adopt the following oplicy:
* Put multiple copies of this large file under our own version control, where each such file corresponds to one Matter release (as in Matter 1.0, Matter 1.2 and so on). If we are talking about the current everything-in-a-single-large-IDL-file status quo, perhaps I'm over-exaggerating the problem of maintaining these up to date under our version control, as they would be just 1:1 copies from the `chip` repo, possibly from different branches of course. So yeah, how about that?

> * how can I split out the "one large IDL" into individual pieces (or do we want one large file and pick from it)?

What if we **don't** split the "one large IDL" A bit `bindgen` like: user always gets _everything_ from the IDL file which is `idl_import`-ed. If only a fraction of that is needed/used by the user - no big deal. The unused stuff will be pruned by the Rust compiler anyway.
 
> * how can I ensure compatibility so changes in CPP does not break matter-rs?

Well I hope one IDL file corresponds to one Matter spec, and when this spec is released, the IDL file no longer sees changes, modulo bug fixes? And then we should copy the Matter 1.0, 1.2 and so on IDLs from the corresponding branches in the `chip` repo, which - I hope - won't be changed once the relevant `chip` branch is released to production?


> * feel uneasy about anything requiring network during the build - I would want some checkout and potentially submodules to be sufficient for compilation (some countries and setups have firewalls preventing generall internet)

Well if it is a single IDL only, then I'm warming up to the idea of keeping a copy (or multiple copies) in our repo, as the maintenance will be very low (we need to update these from time to time; more relevant for the latest IDL file than the older ones, corresponding to older Matter spec releases).

Keeping a copy in our repo - more specifically - in the `src` folder of the IDL parser crate - would also allow us to get rid of your `[env]` hack, if we manage to utilize the `include_str!` macro.

Finally - and over time - we can always provide a variation of the `import_idl` macro, which can pull the IDL from custom locations as an alternative of the `include_str!` embedded I suggested above.

> * idl is the friendliest format we have, however there is also movement to generate xml directly out of spec text, which would be the most accurate. Those are split (https://github.com/project-chip/connectedhomeip/tree/master/data_model/clusters) however not currently fully correct/usable so we may want to hold off on that path (hence the current model includes a idl parser)

I see. I guess we'll solve that when the time comes. As long as we don't try to massage the generated XML ourselves, manually (which I think is my primary and only concern with the IDL approach in the meantime as well), we should be able to figure out an approach which does not result in too much maintenance for us.

Reply:
I could not find a way to get include_str! to work. If I pass that into a macro, I end up with a tokenstream of includestr I believe :( include_str works great and I wanted to figure out how it gets the paths without the `env` bit however I think docs/source say `compiler built-in`.

For the future, we may also want to support custom clusters, which would not reside in the sdk/spec files, so I would like to keep some way to `include_idl` (or whatever follows like `include_dm_xml`) to specify a custom file path.

Reply:
Updated code to be a single large file and allow the macro to have a `clusters = ...` filter.